core/src/mast/merger/mod.rs

```diff
@@ -0,0 +1,429 @@
+
+                MastNode::new_split(if_branch, else_branch, &self.mast_forest)
+                    .expect("SplitNode children should have been mapped to a lower index")
+            },
+            MastNode::Loop(loop_node) => {
+                let body = map_node_id(loop_node.body());
+                MastNode::new_loop(body, &self.mast_forest)
+                    .expect("LoopNode children should have been mapped to a lower index")
+            },
+            MastNode::Call(call_node) => {
+                let callee = map_node_id(call_node.callee());
+                MastNode::new_call(callee, &self.mast_forest)
+                    .expect("CallNode children should have been mapped to a lower index")
+            },
+            // Other nodes are simply copied.
+            MastNode::Block(basic_block_node) => {
+                MastNode::new_basic_block(
+                    basic_block_node.operations().copied().collect(),
+                    // Operation Indices of decorators stay the same while decorator IDs need to be
+                    // mapped.
+                    Some(
+                        basic_block_node
+                            .decorators()
+                            .iter()
+                            .map(|(idx, decorator_id)| match map_decorator_id(decorator_id) {
+                                Ok(mapped_decorator) => Ok((*idx, mapped_decorator)),
+                                Err(err) => Err(err),
+                            })
+                            .collect::<Result<Vec<_>, _>>()?,
+                    ),
+                )
+                .expect("previously valid BasicBlockNode should still be valid")
+            },
+            MastNode::Dyn(_) => MastNode::new_dyn(),
+            MastNode::External(external_node) => MastNode::new_external(external_node.digest()),
+        };
+
+        // Decorators must be handled specially for basic block nodes.
+        // For other node types we can handle it centrally.
+        if !mapped_node.is_basic_block() {
+            mapped_node.set_before_enter(map_decorators(node.before_enter())?);
+            mapped_node.set_after_exit(map_decorators(node.after_exit())?);
+        }
+
+        Ok(mapped_node)
+    }
+
+    // HELPERS
+    // ================================================================================================
+
+    fn lookup_node_by_fingerprint(&self, eq_hash: &EqHash) -> Option<&(EqHash, MastNodeId)> {
+        self.node_id_by_hash.get(&eq_hash.mast_root).and_then(|node_ids| {
+            node_ids.iter().find(|(node_fingerprint, _)| node_fingerprint == eq_hash)
+        })
+    }
+
+    fn lookup_node_by_root(&self, mast_root: &RpoDigest) -> Option<&(EqHash, MastNodeId)> {
+        self.node_id_by_hash.get(mast_root).and_then(|node_ids| node_ids.first())
+    }
+
+    fn lookup_external_node_by_root(&self, fingerprint: &EqHash) -> Option<(EqHash, MastNodeId)> {
```

Comment:
```suggestion
    fn lookup_external_node_by_fingerprint(&self, fingerprint: &EqHash) -> Option<(EqHash, MastNodeId)> {
```

Reply:
Thanks! The function name is actually correct because it does look up by root, it just used the root in the `EqHash`. I changed it to take the MAST root directly instead.